usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -2912,6 +2917,177 @@ static int get_curve_type_from_template(TEMPLATE *tmpl)
     return curve_type;
 }
 
+/* import a AES-XTS key, that is, make a blob for a AES XTS key
+ * that was not created by EP11 hardware, encrypt the key by the wrap key,
+ * unwrap it by the wrap key
+ */
+static CK_RV import_aes_xts_key(STDLL_TokData_t *tokdata, SESSION *sess,
+                                OBJECT *aes_xts_key_obj,
+                                CK_BYTE *blob, size_t *blob_size)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_BYTE cipher[MAX_BLOBSIZE];
+    CK_ULONG clen = sizeof(cipher);
+    CK_BYTE csum[MAX_CSUMSIZE];
+    size_t cslen = sizeof(csum);
+    CK_BYTE iv[AES_BLOCK_SIZE];
+    size_t blob_size2 = *blob_size;
+    CK_MECHANISM mech = { CKM_AES_CBC_PAD, iv, AES_BLOCK_SIZE };
+    CK_RV rc;
+    CK_ATTRIBUTE_PTR p_attrs = NULL;
+    CK_ULONG attrs_len = 0;
+    CK_ATTRIBUTE_PTR new_p_attrs = NULL;
+    CK_ULONG new_attrs_len = 0;
+    CK_ATTRIBUTE *chk_attr = NULL, *attr = NULL;
+    unsigned char *ep11_pin_blob = NULL;
+    CK_ULONG ep11_pin_blob_len = 0;
+    ep11_session_t *ep11_session = (ep11_session_t*) sess->private_data;
+
+    if (ep11tok_is_mechanism_supported(tokdata, CKM_AES_XTS) != CKR_OK) {
+        TRACE_ERROR("%s\n", ock_err(ERR_KEY_TYPE_INCONSISTENT));
+        return CKR_KEY_TYPE_INCONSISTENT;
+    }
+
+    rc = template_attribute_get_non_empty(aes_xts_key_obj->template, CKA_VALUE,
+                                          &attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_VALUE for the key.\n");
+        return rc;
+    }
+
+    /* tell ep11 the attributes the user specified */
+    rc = build_ep11_attrs(tokdata, aes_xts_key_obj->template, &p_attrs,
+                          &attrs_len, CKK_AES, CKO_SECRET_KEY, -1, &mech);
+    if (rc != CKR_OK)
+        goto import_aes_xts_key_end;
+
+    rc = ep11tok_pkey_check_aes_xts(tokdata, p_attrs, attrs_len);
```

Comment:
Function ep11tok_pkey_check_aes_xts() checks if CKM_AES_XTS_KEY_GEN is supported, which is somewhat superfluous for importing. Import checks for CKM_AES_XTS being supported above already (which is correct).

Maybe we can remove the check for CKM_AES_XTS_KEY_GEN being supported from `ep11tok_pkey_check_aes_xts()` and just check for the respective mechanism being supported in the key gen function and import function individually? That way function `ep11tok_pkey_check_aes_xts()` is pretty much superfluous as well, since it only consist of 
```
        if (object_is_extractable(key_obj) ||
            !object_is_pkey_extractable(key_obj) ||
            object_is_attr_bound(key_obj))
....
```
(after the change of my previous comment).

Reply:
Or alternatively pass the mechanism to check to `ep11tok_pkey_check_aes_xts()`, then you can use that function from keygen as well as from import, and you don't need the mechanism check in import separately....